keywrap/pkcs11/keywrapper_pkcs11.go

```diff
@@ -0,0 +1,233 @@
+func getPublickey(label string, p *pkcs11.Ctx, sh pkcs11.SessionHandle) (pub pkcs11.ObjectHandle, err error) {
+	pub, err = findObject(p, sh, pkcs11.CKO_PUBLIC_KEY, label)
+	if err != nil {
+		pub, _, err = generateRSAKeyPair(p, sh, label, true)
+	}
+	return
+}
+
+func findObject(p *pkcs11.Ctx, sh pkcs11.SessionHandle, class uint, label string) (pkcs11.ObjectHandle, error) {
+	template := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, class),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, label),
+	}
+	if err := p.FindObjectsInit(sh, template); err != nil {
+		return 0, errors.Wrap(err, "FindObjectsInit")
+	}
+	obj, _, err := p.FindObjects(sh, 1)
+	if err != nil {
+		return 0, errors.Wrap(err, "FindObjects")
+	}
+	if err := p.FindObjectsFinal(sh); err != nil {
+		return 0, errors.Wrap(err, "FindObjectsFinal")
+	}
+	if len(obj) > 0 {
+		return obj[0], nil
+	}
+	return 0, errors.New("Not Found Object")
+}
+
+func generateRSAKeyPair(p *pkcs11.Ctx, session pkcs11.SessionHandle, tokenLabel string, tokenPersistent bool) (pkcs11.ObjectHandle, pkcs11.ObjectHandle, error) {
+	publicKeyTemplate := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PUBLIC_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, pkcs11.CKK_RSA),
+		pkcs11.NewAttribute(pkcs11.CKA_TOKEN, tokenPersistent),
+		pkcs11.NewAttribute(pkcs11.CKA_VERIFY, true),
+		pkcs11.NewAttribute(pkcs11.CKA_PUBLIC_EXPONENT, []byte{1, 0, 1}),
+		pkcs11.NewAttribute(pkcs11.CKA_MODULUS_BITS, 2048),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, tokenLabel),
+	}
+	privateKeyTemplate := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_TOKEN, tokenPersistent),
+		pkcs11.NewAttribute(pkcs11.CKA_SIGN, true),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, tokenLabel),
+		pkcs11.NewAttribute(pkcs11.CKA_SENSITIVE, true),
+		pkcs11.NewAttribute(pkcs11.CKA_EXTRACTABLE, true),
+	}
+	pbk, pvk, err := p.GenerateKeyPair(session,
+		[]*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_RSA_PKCS_KEY_PAIR_GEN, nil)},
+		publicKeyTemplate, privateKeyTemplate)
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "Failed to generate keypair")
+	}
+
+	return pbk, pvk, nil
+}
+
+// loginDevice login Device
+func loginDevice(modules [][]byte, pins [][]byte) (ctx *pkcs11.Ctx, session pkcs11.SessionHandle, err error) {
+	if len(pins) == 0 {
+		return nil, 0, errors.New("Need input module pin")
+	}
+	pin := string(pins[0])
+	pin = strings.TrimSpace(pin)
+
+	if len(modules) > 1 {
+		return nil, 0, errors.New("Just support single module")
+	}
+	module := string(modules[0])
+	module = strings.TrimSpace(module)
+	ctx = pkcs11.New(module)
+	if ctx == nil {
+		return nil, 0, errors.New("Please check Module path, input is: " + module)
+	}
+
+	err = ctx.Initialize()
+	if err != nil {
+		p11Err := err.(pkcs11.Error)
+		if p11Err != pkcs11.CKR_CRYPTOKI_ALREADY_INITIALIZED {
+			return nil, 0, errors.Wrap(err, "Device Initialize failed")
+		}
+	}
+	slots, err := ctx.GetSlotList(true)
```

Comment:
note for myself: Future enhancement for specifying slot number

Reply:
added.